ecommerce/serializers.py

```diff
@@ -107,6 +118,120 @@ class Meta:
+
+    class Meta:
+        fields = ("id", "product")
+        model = models.Line
+
+
+class OrderSerializer(serializers.ModelSerializer):
+    """ Order Serializer """
+
+    name = serializers.SerializerMethodField()
+    close_date = serializers.SerializerMethodField(allow_null=True)
+    amount = serializers.SerializerMethodField()
+    discount_amount = serializers.SerializerMethodField()
+    coupon_code = serializers.SerializerMethodField(allow_null=True)
+    company = serializers.SerializerMethodField(allow_null=True)
+    line_items = serializers.SerializerMethodField()
+    b2b = serializers.SerializerMethodField()
+
+    def get_name(self, instance):
+        """ Return the deal name """
+        return f"XPRO-ORDER-{instance.id}"
+
+    def get_close_date(self, instance):
+        """ Return the updated_on date (as a timestamp in milliseconds) if fulfilled """
+        if instance.status == models.Order.FULFILLED:
+            return int(instance.updated_on.timestamp() * 100)
+
+    def get_amount(self, instance):
+        """ Get the amount paid after discount """
+        product_version = ProductVersion.objects.filter(
+            id__in=instance.lines.values_list("product_version", flat=True)
+        ).first()
+        coupon_version = CouponVersion.objects.filter(
+            id__in=CouponRedemption.objects.filter(order=instance).values_list(
+                "coupon_version__id", flat=True
+            )
+        ).first()
+        return get_product_version_price_with_discount(
+            coupon_version=coupon_version, product_version=product_version
+        ).to_eng_string()
+
+    def get_discount_amount(self, instance):
+        """ Get the discount amount if any """
+        coupon_version = CouponVersion.objects.filter(
+            id__in=CouponRedemption.objects.filter(order=instance).values_list(
+                "coupon_version__id", flat=True
+            )
+        ).first()
+        if not coupon_version:
+            return "0.00"
+
+        product_version = ProductVersion.objects.filter(
+            id__in=instance.lines.values_list("product_version", flat=True)
+        ).first()
+
+        return round_half_up(
+            coupon_version.payment_version.amount * product_version.price
+        ).to_eng_string()
+
+    def get_line_items(self, instance):
+        """ Get the list of line items"""
+        return [LineSerializer(line).data for line in instance.lines.all()]
+
+    def get_company(self, instance):
+        """ Get the company id if any """
+        redemption = CouponRedemption.objects.filter(order=instance).last()
+        if redemption:
+            company = redemption.coupon_version.payment_version.company
+            if company:
+                return company.id
+
+    def get_coupon_code(self, instance):
+        """ Get the coupon code used for the order if any """
+        redemption = CouponRedemption.objects.filter(order=instance).first()
```

Comment:
A few things:

- Should only query for `CouponRedemption.objects.filter(order=instance).first()` once per serialization of this object. Same thing with the `ProductVersion` queries going on.
- I see on line 194 we do `last()` and here we do `first()`, is there a reason for these discrepancies? Or should this even just be a `get(order=instance)`?

Reply:
There should be only one query result (1 redemption, 1 coupon version, 1 product version).